components/ui/data-table.tsx

```diff
@@ -0,0 +1,121 @@
+'use client'
+
+import {
+  ColumnDef,
+  flexRender,
+  getCoreRowModel,
+  getPaginationRowModel,
+  useReactTable,
+} from "@tanstack/react-table"
+
+import {
+  Table,
+  TableBody,
+  TableCell,
+  TableHead,
+  TableHeader,
+  TableRow,
+} from "@/components/ui/table"
+import { Button } from "./button"
+import { useEffect, useState } from "react"
+import { LocalApi } from "@/lib/sdk/local"
+import { useReferences } from "@/hooks/useReferences"
+
+interface DataTableProps<TData, TValue> {
+  columns: ColumnDef<TData, TValue>[]
+}
+
+export function DataTable<TData, TValue>({
+  columns,
+}: DataTableProps<TData, TValue>) {
+  const { list, hasNextPage } = useReferences()
+  const [internalData, setInternalData] = useState(list)
+  const table = useReactTable({
+    data: internalData,
+    columns,
+    getCoreRowModel: getCoreRowModel(),
+    getPaginationRowModel: getPaginationRowModel(),
+  })
+  const { pageIndex } = table.getState().pagination
+  const api = new LocalApi()
+
+  useEffect(() => {
+    const pageCount = table.getPageCount()
+    async function fetchMoreData() {
+      if (pageIndex === pageCount - 2 && hasNextPage) {
+        const { references } = await api.getReferencesData(undefined, pageIndex)
```

Comment:
We shouldn't use this sdk endpoint direct on the component.
The whole ideia of have a dedicated hook to deal with the consumption of our references list is delegate to it the progressive request logic

Reply:
Correct. Refactoring it!